readme.txt
$Id: readme.txt,v 1.3 2011/04/1, 05:05:06 wolfpaulus Exp $

Introduction
============

See the web site http://www.swixml.org


Installing the build tools
==========================

The SWIXML build system is based on Ant, which is a Java building tool
originally developed for the Jakarta Tomcat project but now used in many other
Apache projects and extended by many developers.

Ant is a little but very handy tool that uses a build file written in XML
(build.xml) as building instructions. For more information refer to
"http://ant.apache.org".

The only thing that you have to make sure of is that the "JAVA_HOME"
environment property is set to match the top level directory containing the
JVM you want to use. 

Building instructions
=====================

Ok, let's build the code. First, make sure your current working directory is
where the build.xml file is located. Then type

  ant 
- or -
  ant distribution

if everything is right and all the required packages are visible, this action
will generate a file called "swixml.jar" in the "./build" directory. Note, that
if you do further development, compilation time is reduced since Ant is able
to detect which files have changed and recompile them as needed.

If something went wrong, go to the Swixml forum, available at the swixml web site.at

  http://www.swixml.org/forum/


Build targets
=============

The build system is not only responsible for compiling SWIXML into a jar file,
but is also responsible for creating the HTML documentation in the form of
javadocs.

These are the meaningful targets for this build file:

 - package [default] -> creates ./build/swixml.jar
 - compile -> compiles the source code
 - samples -> compiles example code
 - javadoc -> generates the API documentation in ./build/javadocs
 - distribution -> builds everything, incl. jar, javadoc, tagdoc, xsd, etc. etc.
 - clean -> restores the distribution to its original and clean state

For example, to build the samples, type

build samples

To learn the details of what each target does, read the build.xml file.  
It is quite understandable.

Best,
Wolf Paulus and the Swixml Dev. Team.



changes.txt

* * * * * * 1.6 (tag: swixml_161) from 1.5 (tag: swixml_161) * * * * *
Swixml 1.6:
Update buildscripts
Added TagDocs source, generating the tag documentation in HTML via introspection 
Update 3rd party libs

Prepare for move to GITHUB

* * * * * * 1.6 (tag: swixml_151) from 1.5 (tag: swixml_146) * * * * *
Swixml 1.6  beta 1:
PrimitiveConverter anow also implement the ScrollPaneConstants interface,
making its constants available to the Swixml descriptor.
JGoodies Formlayout support thanks to Karl Tauber of JFormDesigner.
Updated 3rd party libraries: Qaqua, Substance, ant, forms, j2h
Updated samples and Javadoc.
Updated Parser code.

* * * * * * 1.5 (tag: swixml_149) from 1.5 (tag: swixml_146) * * * * *
Added Swixml XML-Schema Generator

Fixed BorderConverter to accept a more complex declaration of a
TitledBorder, which can now be declared like so:
border="TitledBorder(TitledBorder, myTitle, TitledBorder.CENTER,
TitledBorder.BELOW_BOTTOM, VERDANA-BOLD-18, blue)"
Added TitledBorder's constants to the PrimitiveConverter

Fixed class return type in the Font Converter.
Supported plaf attributes on Macs.
Cleaned up ConverterLibrary.
Removed <i>final</a> from SwingEngine's static DEBUG member

Added CustomTag sample which uses the JCalendar, Qaqua, and Substance
3rd party jars, all added to the lib folder.

Wrapping client methods into AbstactActions:
If a client's method should be used as an Action but the client is not
intersted in the ActionEvent that caused it, there may not a reason to
imlement an Action but rather use a simple operation like:
public void myOperation() { .. }
If a client's method name is found in the action attribute's value and
an Action with this name doesn't exist, then the method is wrapped
into an AbstactAction and mapped.
 
Wrap the JScrollPane to override the setBackground method. Setting the
background of a scrollpane can be accomplished by setting the
background color of the viewport, via
scrollPane.getViewport().setBackground().

Localizer class is now localizing comma sep. string lists, which now
for instance allows the TabbedPane's titles attribute to be used
together with localization.

* * * * * * 1.5 (tag: swixml_146) from 1.5 (tag: swixml_145) * * * * *
Switched VCS from CVS to SVN
Replaced 3rd party libs with newer version: ui.jar, ant.jar, junit.jar
 
* * * * * * 1.5 (tag: swixml_145) from 1.4 (tag: swixml_144) * * * * *
Fixed AccessControlException thrown when run in applet or otherwise
restricted environment.

* * * * * * 1.4 (tag: swixml_141) from 1.3 (tag: swixml_144) * * * * *
Parser now adds a JMenuBar into its parent-container,
in case the parent doesn't provide a setJMenuBar(..) method.
Automatic field mapping now works with private field too.
Transient fields however, are not mapped.
ButtonGroups with an id attribute are now mapped like Components.
Added lib/junit.jar into build.
Replaced lib/jdom.jar with Jason Hunter's 1.0 release.

* * * * * * 1.3 (tag: swixml_139) from 1.2 (tag: swixml_141) * * * * *
Actions instantiated through initclass are now working on MacOS.

The actionPerformed of all MacOS-actions does not get a null reference
as ActionEvent anymore. Instead the original (ApplicationEvent)-source
is used. Further, the ActionEvent gets a (somewhat) useful
actionCommand String specifying the called function, e.g. for the
Action mapped to MacOS-File-Open-Command the actionCommand
Parser.ATTR_MACOS_OPENFILE.

Thanks to Volker Bublitz for help and testing environment.

* * * * * * 1.2 (tag: swixml_138) from 1.1 (tag: swixml_139) * * * * *
Changed Mac OS X detection like recommended in Apple's Technical Note TN2110

To avoid the problem with BufferedReader switching to the default
Platform-Encoding,
BufferedReaders are not used anymore when reading XML descriptors.
i.e. DataInputStreams are implicitly localized when wrapped into an appropriate
BufferedReader.

Deprecated SwingEngine Ctor: SwingEngine(ClassLoader cl, final String resource),
which is not useful but confusing, i.e. a ResourceBundle can only be set after
the XML descriptor has been rendered, when changing the CL however, one often
wants to provide a new rb as well.

Fixed the Localizer class' setResourceBundle() method to work also in a remoting
environment.

* * * * * * 1.1 (tag: swixml_138) from 1.0.1 (tag: swixml_137) * * * * *
Added example HelloMac, showing some of the Mac OS X specifics
Updated TagDocumentation/Custom Swixml Attributes with MAC_OS examples
Updated TagDocumentation to link to Java 2 Platform Std Ed. v1.4.2
Updated lib/jdom.jar library to BETA-10 RC1.
Updated lib/ant.jar to ant 1.6.1
Updated lib/ui.jar to apple jre 1.4.2
Updated build.xml file to work on the MAC
Updated license file.

Added guessSetter method to Factory Interface and implementations.
Added guessSetter method to TagLibrary class.
The new methods are used during parsing to determine which method to used to
set a property and case now doesn't matter anymore for attribute names.
I.e. a JLabels' MaximumSize property could be set with a label tag like this:
<label MaximumSize="123"/>. However, <label Maximumsize="123"/> didn't work before.

BUG FIXES
---------
Fixed logical error in method SwingEngine.setActionListener, which prevented 
ActionListeners being assigned to MenuItems.

* * * * * * 1.0.1 (tag: swixml_136) from 1.0 (tag: swixml_135) * * * * *
Added ClassLoader property to the SwingEngine to allow remoting.
The classloader can be set through manually or through a constructor or if not set,
the SwingEngines ClassLoader is used instead. In any case, the classloader is
passed to the Localizer and used by the Parser and the IconImageConverter.

Implemented a RectangleConverter and added to ConverterLib.,
This has a nice side effect; it enables methods like setBounds(), i.e. for
JInternalFrames. Therefore, a Bounds attribute is now supported for the internalframe tag.

Added more sample code to smaples package, i.e. Layout sample showing some of the

src/org/swixml/converters/BorderConverter.java
/*--
 $Id: BorderConverter.java,v 1.1 2004/03/01 07:55:58 wolfpaulus Exp $

 Copyright (C) 2003-2007 Wolf Paulus.
 All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:

 1. Redistributions of source code must retain the above copyright
 notice, this list of conditions, and the following disclaimer.

 2. Redistributions in binary form must reproduce the above copyright
 notice, this list of conditions, and the disclaimer that follows
 these conditions in the documentation and/or other materials provided
 with the distribution.

 3. The end-user documentation included with the redistribution,
 if any, must include the following acknowledgment:
        "This product includes software developed by the
         SWIXML Project (http://www.swixml.org/)."
 Alternately, this acknowledgment may appear in the software itself,
 if and wherever such third-party acknowledgments normally appear.

 4. The name "Swixml" must not be used to endorse or promote products
 derived from this software without prior written permission. For
 written permission, please contact <info_AT_swixml_DOT_org>

 5. Products derived from this software may not be called "Swixml",
 nor may "Swixml" appear in their name, without prior written
 permission from the Swixml Project Management.

 THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED
 WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 DISCLAIMED.  IN NO EVENT SHALL THE SWIXML PROJECT OR ITS
 CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 SUCH DAMAGE.
 ====================================================================

 This software consists of voluntary contributions made by many
 individuals on behalf of the Swixml Project and was originally
 created by Wolf Paulus <wolf_AT_swixml_DOT_org>. For more information
 on the Swixml Project, please see <http://www.swixml.org/>.
*/

package org.swixml.converters;

import org.jdom.Attribute;
import org.swixml.Converter;
import org.swixml.ConverterLibrary;
import org.swixml.Localizer;
import org.swixml.SwingEngine;

import javax.swing.*;
import javax.swing.border.Border;
import java.lang.reflect.Method;
import java.util.StringTokenizer;

/**
 * The <code>BorderConverter</code> class defines a converter that creates Border objects based on a provided String.
 * The BorderConverter internally uses the <code>javax.swing.BorderFactory</code> and its static <i>create</i>.. methods
 * to instatiate differnt kinds of borders, based on the given String.<br>
 * Additional parameters to create a border need to be comma separated and enclosed in parentheses.<br><br>
 * Parameter types get converted through the <code>ConverterLibrary</code>. Therefore, only parameter classes are supported that
 * have registered  converters in the ConverLibrary.
 * <h3>Examples for Valid XML attribute notations:</h3>
 * <pre>
 * <ul>
 * <li>border="MatteBorder(4,4,4,4,red)"</li>
 * <li>border="EmptyBorder(5,5,5,5)"</li>
 * <li>border="TitledBorder(My Title)"</li>
 * <li>border="RaisedBevelBorder"</li>
 * <li>border="TitledBorder(TitledBorder, myTitle, TitledBorder.CENTER, TitledBorder.BELOW_BOTTOM, VERDANA-BOLD-18, blue)"</li>
 * </ul>
 * </pre>
 *
 * @author <a href="mailto:wolf@wolfpaulus.com">Wolf Paulus</a>
 * @version $Revision: 1.1 $
 * @see javax.swing.BorderFactory
 * @see javax.swing.border.AbstractBorder
 * @see org.swixml.ConverterLibrary
 */
public class BorderConverter implements Converter {
  /**
   * converter's return type
   */
  public static final Class TEMPLATE = Border.class;

  /**
   * all methods the BorderFactory provides
   */
  private static final Method[] METHODS = BorderFactory.class.getMethods();

  /**
   * Returns a <code>javax.swing Border</code>
   *
   * @param type <code>Class</code>  not used
   * @param attr <code>Attribute</code> value needs to provide Border type name and optional parameter
   * @return <code>Object</code> runtime type is subclass of <code>AbstractBorder</code>
   */
  public Object convert(final Class type, final Attribute attr, Localizer localizer) {
    Border border = null;
    StringTokenizer st = new StringTokenizer(attr.getValue(), "(,)"); // border type + parameters
    int n = st.countTokens() - 1; // number of parameter to create a border
    String borderType = st.nextToken().trim();
    Method method = null;
    ConverterLibrary cvtlib = ConverterLibrary.getInstance();
    //
    // Special case for single parameter construction, give priority to String Type
    //
    if (n == 0) {
      try {
        method = BorderFactory.class.getMethod("create" + borderType);

      } catch (NoSuchMethodException e) {
        // intent. empty
      }
      if (method == null) { // try with empty string
        n = 1;
        st = new StringTokenizer(" ", "(,)");
      }
    }
    if (n == 1) {
      try {
        method = BorderFactory.class.getMethod("create" + borderType, new Class[]{String.class});
      } catch (NoSuchMethodException e) {
        //  no need to do anything here.
      }
    }
    for (int i = 0; method == null && i < METHODS.length; i++) {
      if (METHODS[i].getParameterTypes().length == n && METHODS[i].getName().endsWith(borderType)) {
        method = METHODS[i];

        for (int j = 0; j < method.getParameterTypes().length; j++) {
          if (String.class.equals(method.getParameterTypes()[j])) {
            continue;
          }
          if (null == cvtlib.getConverter(method.getParameterTypes()[j])) {
            method = null;
            break;
          }
        }
      }
    }
    try {
      Object[] args = new Object[n];
      for (int i = 0; i < n; i++) { // fill argument array
        Converter converter = cvtlib.getConverter(method.getParameterTypes()[i]);
        Attribute attrib = new Attribute(String.class.equals(converter.convertsTo()) ? "title" : "NA", st.nextToken().trim(), Attribute.CDATA_TYPE);
        if (converter != null) {
          args[i] = converter.convert(method.getParameterTypes()[i], attrib, localizer);
        } else {
          args[i] = attrib.getValue();
        }
      }
      border = (Border) method.invoke(null, args);
    } catch (Exception e) {
      if (SwingEngine.DEBUG_MODE)
        System.err.println("Couldn't create border, " + attr.getValue() + "\n" + e.getMessage());
    }
    return border;
  }

  /**
   * A <code>Converters</code> conversTo method informs about the Class type the converter
   * is returning when its <code>convert</code> method is called
   *
   * @return <code>Class</code> - the Class the converter is returning when its convert method is called
   */
  public Class convertsTo() {
    return TEMPLATE;
  }
}
